elixir/lib/contracts.ex

```diff
@@ -14,9 +16,63 @@ defmodule Trento.Contracts do
   Encode and wrap an event struct to a protobuf CloudEvent.
   """
   @spec to_event(struct(), Keyword.t()) :: binary()
-  def to_event(%mod{} = struct, opts \\ []) do
+  def to_event(%{} = struct, opts \\ []) do
+    struct
+    |> build_cloud_event(opts)
+    |> CloudEvents.CloudEvent.encode()
+  end
+
+  @doc """
+  Encode and wrap an event struct to a signed protobuf Cloudevent.
+  Accepts the private key as a pem file content.
+  Accepts the same options as to_event/2
+  """
+  @spec to_signed_event(struct(), binary(), Keyword.t()) :: binary()
+  def to_signed_event(struct, pem_private_key, opts \\ []) do
+    struct
+    |> build_cloud_event(opts)
+    |> add_signature(pem_private_key)
+    |> CloudEvents.CloudEvent.encode()
+  end
+
+  @doc """
+  Decode and unwrap a protobuf CloudEvent to an event struct.
+  """
+  @spec from_event(binary()) ::
+          {:ok, struct()}
+          | {:error, :decoding_error}
+          | {:error, :invalid_envelope}
+          | {:error, :unknown_event}
+          | {:error, :event_expired}
+  def from_event(value) do
+    with {:ok, type, cloud_event} <- decode_cloud_event(value),
+         {:ok, cloud_event} <- verify_event_validity(cloud_event) do
+      decode_trento_event(type, cloud_event)
+    end
+  end
+
+  @doc """
+  Decode and unwrap a signed protobuf CloudEvent to an event struct.
+  """
+  @spec from_signed_event(binary(), binary()) ::
+          {:ok, struct()}
+          | {:error, :decoding_error}
+          | {:error, :invalid_envelope}
+          | {:error, :unknown_event}
+          | {:error, :event_expired}
+          | {:error, :invalid_event_signature}
+  def from_signed_event(value, public_key) do
+    with {:ok, event_type, event_data} <- decode_cloud_event(value),
+         {:ok, event_data} <- verify_event_signature(event_data, public_key),
+         {:ok, event_data} <- verify_event_validity(event_data) do
+      decode_trento_event(event_type, event_data)
+    end
+  end
+
+  defp build_cloud_event(%mod{} = struct, opts) do
     id = Keyword.get(opts, :id, UUID.uuid4())
     source = Keyword.get(opts, :source, "trento")
+    validity_in_seconds = Keyword.get(opts, :validity_in_seconds, @default_event_validity)
```

Comment:
nitpick: maybe some other name like `valid_for` or `ttl`?
I mean, the `in_seconds` looks a bit strange. Maybe just documenting would be enough.
I have not seen many of these things in elixir though